File: go-ut-bench/datasets/cpp/cpp_code_files_self_contained/interface_mock/interface_mock_000.cpp
#include <string>
#include <stdexcept>

class DataSource {
public:
    virtual std::string read() = 0;
    virtual bool write(const std::string& data) = 0;
    virtual ~DataSource() {}
};

class FileDataSource : public DataSource {
private:
    std::string filepath;
    std::string data;
    
public:
    FileDataSource(const std::string& path) : filepath(path), data("") {}
    
    std::string read() override {
        return data;
    }
    
    bool write(const std::string& newData) override {
        data = newData;
        return true;
    }
};

class MockDataSource : public DataSource {
private:
    std::string data;
    
public:
    MockDataSource() : data("mock_data") {}
    
    std::string read() override {
        return data;
    }
    
    bool write(const std::string& newData) override {
        data = newData;
        return true;
    }
};

class DataProcessor {
private:
    DataSource* source;
    
public:
    DataProcessor(DataSource* src) : source(src) {}
    
    std::string process() {
        std::string data = source->read();
        std::string result;
        for (char c : data) {
            if (c >= 'a' && c <= 'z') {
                result += static_cast<char>(c - 32);
            } else {
                result += c;
            }
        }
        return result;
    }
    
    bool save(const std::string& data) {
        return source->write(data);
    }
};

File: go-ut-bench/datasets/go/go_code_files_self_contained/interface_mock/interface_mock_000.go
package main

type DataSource interface {
	Read() string
	Write(data string) bool
}

type FileDataSource struct {
	filepath string
	data     string
}

func NewFileDataSource(filepath string) *FileDataSource {
	return &FileDataSource{filepath: filepath, data: ""}
}

func (fds *FileDataSource) Read() string {
	return fds.data
}

func (fds *FileDataSource) Write(data string) bool {
	fds.data = data
	return true
}

type MockDataSource struct {
	data string
}

func NewMockDataSource() *MockDataSource {
	return &MockDataSource{data: "mock_data"}
}

func (mds *MockDataSource) Read() string {
	return mds.data
}

func (mds *MockDataSource) Write(data string) bool {
	mds.data = data
	return true
}

type DataProcessor struct {
	source DataSource
}

func NewDataProcessor(source DataSource) *DataProcessor {
	return &DataProcessor{source: source}
}

func (dp *DataProcessor) Process() string {
	data := dp.source.Read()
	result := ""
	for _, c := range data {
		if c >= 'a' && c <= 'z' {
			result += string(c - 32)
		} else {
			result += string(c)
		}
	}
	return result
}

func (dp *DataProcessor) Save(data string) bool {
	return dp.source.Write(data)
}